protocol/blockchain/transaction/transaction-signing.md

```diff
@@ -127,7 +127,8 @@ The resulting script is what is included in the signature preimage.
 If the hash type is `SIGHASH_NONE` then the hash should be all `0x00` bytes.
 
 If hash type is `SIGHASH_SINGLE` then only the output with the same index as the input being signed is included.
-If no such output exists (i.e. there are fewer outputs than the index of the input to be signed), this is again all `0x00` bytes.
+If no such output exists (i.e. there are fewer outputs than the index of the input to be signed), then due to a bug inherited from Bitcoin Core the top signature hash will be constant `0x0100000000000000000000000000000000000000000000000000000000000000`.
+Signers should NEVER use their address key to sign a "message" equal to that constant as that would result in giving up access to ALL UTXOs belonging to the key on ALL blockchains that inherited the bug.
```

Comment:
holy shit. What a pitfall ripe for abuse.

Reply:
I learned of this only today when Ruben Somsen brought it up on Telegram, but apparently it's been known since 2014.
I found this in BU codebase:

```
            //  nOut out of range
            // IMPORTANT NOTICE:
            // Returning one from SignatureHash..() to signal error conditions is a kludge that
            // is also breaking the ECDSA assumption that only cryptographic hashes are signed. The special value
            // returned here is, however, due to further omissions in CheckSig, part of the pre-BCH
            // consensus rule set and needs to be left as-is.
            // See also: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2014-November/006878.html
```

https://gitlab.com/bitcoinunlimited/BCHUnlimited/-/blob/dev/src/script/sigcommon.cpp#L186

It's funny that over the course of last year I read that [docs page](https://documentation.cash/protocol/blockchain/transaction/transaction-signing) multiple times, even read the [section below that describes the bug on Bitcoin Core](https://documentation.cash/protocol/blockchain/transaction/transaction-signing#bitcoin-core-signatures), but didn't get triggered to investigate further, although I remember wondering "why is this here?". This PR should make it clear what it is about :)

I don't know who originally wrote this page of our docs, but it seems that it was written as if the bug was fixed on Bitcoin Cash and the SIGHASH_SINGLE behaves as it logically should (in case of non-existant output then just the outputs hash should be a 32b zero, and the preimage should still have the input data).

Reply:
I'm fairly certain that the original documentation is correct.  When `SIGHASH_FORKID` is set the BIP-143 algorithm (from 2016) is used instead, which is what is being described here.

The BU comment referenced above is in the method `SignatureHashLegacy`.  The only only references to `SIGNATURE_HASH_ERROR` are in that method as well.  The method `SignatureHashBitcoinCash` does not appear to have any similar logic either.  I also confirmed that [the Bitcoin Verde transaction signing logic](https://github.com/SoftwareVerde/bitcoin-verde/blob/master/src/main/java/com/softwareverde/bitcoin/transaction/signer/TransactionSigner.java) follows the same pattern.

It's possible some additional clarification could be helpful to avoid this kind of confusion in the future, but I believe the current state of the documentation does match what currently happens on the network.

Reply:
Oh that's right, sorry for the false alarm then. I looked through all our upgrade specs but didn't know what to look for, expected an explicit reference or the bug but there's none. Here's the part about signsture digest in the UAHF specification:

>REQ-6-3 (use adapted BIP143 hash algorithm for protected transactions)

>Once the fork has activated, any transaction that has bit 6 set in its hash type shall have its signature hash computed using a minimally revised form of the transaction digest algorithm specified in BIP143.

>RATIONALE: see Motivation section of BIP143 [2].

>NOTE 1: refer to [3] for the specificaton of the revised transaction digest based on BIP143. Revisions were made to account for non-Segwit deployment.

Ok, I'll investigate further and see how to improve this PR to make it more clear.